### ISSUE TITLE
dotnet restore --no-dependencies is loading asset files from upstream project references in the latest .Net 6 preview 

### ISSUE THREAD

**Issue by nkolev92** [dotnet/sdk#20556]
@breidikl commented on [Wed, 11 Aug 2021 21:05:08 GMT](https://github.com/NuGet/Home/issues/11147)
### NuGet Product Used

dotnet.exe

### Product Version

6.0.0-preview.7.21377.19

### Worked before?

It works in 5.x version and also worked in 6 preview 4

### Repro Steps & Context

We have a distributed build that cds to the directory for each csproj and runs the following dotnet command on the machine that will be building that node:

`dotnet.exe restore --use-lock-file --no-dependencies /p:ImportProjectExtensionProps=false /p:ImportProjectExtensionTargets=false /nodeReuse:false /maxcpucount:1 --lock-file-path packages.lock.json --locked-mode `

At some point after .Net 6 preview 4, this started reading the project.assets.json files from the upstream project references, which causes problems with our build distribution model.

**Comment by nkolev92** [comment]
@erdembayar commented on [Sat, 14 Aug 2021 01:35:55 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-898789730)
@breidikl 
Thank you for filing this issue. Could you be able to provide sample solution/project exhibit this issue? 
We need to repro the issue and investigate.

**Comment by nkolev92** [comment]
@breidikl commented on [Tue, 17 Aug 2021 00:19:42 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-899901409)
Here's a more minimalistic repro with projects (netcoreconsoleapp and netcoreclasslibrary as sibling directories each containing a csproj).

netcoreconsoleapp.csproj contains:

```
<Project Sdk="Office.Build.Before;Microsoft.NET.Sdk;Office.Build.After">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\NetCoreClassLibrary\NetCoreClassLibrary.csproj" />
  </ItemGroup>

</Project>
```
netcoreclasslibary.csproj  contains:

```
<Project Sdk="Office.Build.Before;Microsoft.NET.Sdk;Office.Build.After">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
    <PackageReference Include="protobuf-net" Version="3.0.73" />
  </ItemGroup>

</Project>
```
Each has a minimil cs file so that the project will build.  i.e. netcoreconsoleapp.cs is as follows:

```
using System;

namespace NetCoreConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}
```
and netcoreclasslibrary.cs is as follows:

```
using System;

namespace NetCoreClassLibrary
{
    public class Class1
    {
    }
}

```
The lock file for netcoreconsoleapp is as follows:

```
{
  "version": 1,
  "dependencies": {
    ".NETCoreApp,Version=v5.0": {
      "Newtonsoft.Json": {
        "type": "Transitive",
        "resolved": "12.0.3",
        "contentHash": "6mgjfnRB4jKMlzHSl+VD+oUc1IebOZabkbyWj2RiTgWwYPPuaK1H97G1sHqGwPlS5npiF5Q0OrxN1wni2n5QWg=="
      },
      "protobuf-net": {
        "type": "Transitive",
        "resolved": "3.0.73",
        "contentHash": "rcqDbIs7bmsctQk7Jo0zAI5BBDyjlKqzGqCHbtlHKRDyJ5AFPBClS3G5KvITS52hSoobUTWW5NsUmB7Dt/0iaA==",
        "dependencies": {
          "protobuf-net.Core": "3.0.73"
        }
      },
      "protobuf-net.Core": {
        "type": "Transitive",
        "resolved": "3.0.73",
        "contentHash": "UKQlQY4STocSar2WoGA2mJegx3E4rTOa+ur2O8YKmhpJII/VfLILkVAF7Cyko2qljFAZ0CuOTy5uNu5JE/ZQ2A=="
      },
      "netcoreclasslibrary": {
        "type": "Project",
        "dependencies": {
          "Newtonsoft.Json": "12.0.3",
          "protobuf-net": "3.0.73"
        }
      }
    }
  }
}
```

Running dotnet.exe restore --use-lock-file --no-dependencies /p:ImportProjectExtensionProps=false /p:ImportProjectExtensionTargets=false /nodeReuse:false /maxcpucount:1 --lock-file-path packages.lock.json --locked-mode in the directory with netcoreconsoleapp then reads project.assets.json produced by the restore for netcoreclasslibrary (which was run as a previous step). 

We use BuildXL for running the commands in our environment, which flags the access when using the latest .Net 6 preview (as an error condition due to not being declared in our distribution graph).  That did not happen when using 6.0.100-preview.4.21255.9 (or any 5x version).

**Comment by nkolev92** [comment]
@nkolev92 commented on [Mon, 23 Aug 2021 17:33:03 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-903972569)
Hey @breidikl 

I can't really think of a reason why restore would be accessing the assets file, so I have to imagine it's some other component. 

I followed your instructions, and the only thing I couldn't get was the office SDK. 

1. I created the setup matching your description, with the exclusion of the office SDKs.
1. Restored both projects. 
1. Checked the last access time of the assets file. 
1. Ran the command to restore that project only with no dependencies 
1. Checked the access time again.

```
C:\Code\Temp\11147 [master]> $(gci .\NetCoreClassLibrary\obj\project.assets.json).LastAccessTime

Monday, August 23, 2021 10:26:53 AM


C:\Code\Temp\11147 [master]> cd .\netcoreconsoleapp\
C:\Code\Temp\11147\netcoreconsoleapp [master]> dotnet.exe restore  --use-lock-file --no-dependencies /p:ImportProjectExtensionProps=false /p:ImportProjectExtensionTargets=false /nodeReuse:false /maxcpucount:1 --lock-file-path packages.lock.json --locked-mode
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Code\Temp\11147\netcoreconsoleapp [master]> cd ..
C:\Code\Temp\11147 [master]> $(gci .\NetCoreClassLibrary\obj\project.assets.json).LastAccessTime

Monday, August 23, 2021 10:26:53 AM


C:\Code\Temp\11147 [master]>
```

I am using `6.0.100-preview.7.21379.14` to test this. 

Can you determine which process is accessing the assets file in question?

**Comment by nkolev92** [comment]
@breidikl commented on [Mon, 23 Aug 2021 21:59:20 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-904161728)
BuildXL records the access as being by dotnet.exe.  I confirmed that the issue started between .Net 6 preview 4 and preview 5, though it still repros in preview 7.  

I'm actually out of town this week, though I sent the latest response to my team to help with additional information while I'm out.

**Comment by nkolev92** [comment]
@nkolev92 commented on [Tue, 24 Aug 2021 00:20:12 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-904223677)
@ 
Author
breidikl linked me the source that can be used to download these packages, I ended up downloading 0.0.10 for the before and 0.0.2 for the after package (since I imagine there's a global.json in the repo somewhere). 

I attempted it again and I still don't see anything happening due to the restore command. 

```
C:\Code\Temp\11147 [master]> dotnet restore .\netcoreconsoleapp\
  Determining projects to restore...
  Restored C:\Code\Temp\11147\netcoreconsoleapp\netcoreconsoleapp.csproj (in 316 ms).
  Restored C:\Code\Temp\11147\NetCoreClassLibrary\NetCoreClassLibrary.csproj (in 316 ms).
C:\Code\Temp\11147 [master]> $(gci .\NetCoreClassLibrary\obj\project.assets.json).LastAccessTime

Monday, August 23, 2021 5:19:01 PM


C:\Code\Temp\11147 [master]> dotnet.exe restore  .\netcoreconsoleapp\ --no-dependencies /p:ImportProjectExtensionProps=false /p:ImportProjectExtensionTargets=false /nodeReuse:false /maxcpucount:1 --locked-mode
  Determining projects to restore...
  Restored C:\Code\Temp\11147\netcoreconsoleapp\netcoreconsoleapp.csproj (in 290 ms).
C:\Code\Temp\11147 [master]> $(gci .\NetCoreClassLibrary\obj\project.assets.json).LastAccessTime

Monday, August 23, 2021 5:19:01 PM
```

**Comment by nkolev92** [comment]
@dsrini1 commented on [Wed, 25 Aug 2021 04:07:10 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-905165641)
Hi @nkolev92,
I work with @breidikl and I have got the repro of this issue. The actual violation is not the read but probe. The upstream project's project.assets.json file is being probed while running dotnet restore. 

I was able to confirm the probe by dotnet.exe, using the [process monitor tool](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) while running the dotnet.exe restore. The dotnet.exe performs "QueryOpen" operation alone on that project.assets.json file.

**Comment by nkolev92** [comment]
@nkolev92 commented on [Wed, 25 Aug 2021 17:46:39 GMT](https://github.com/NuGet/Home/issues/11147#issuecomment-905744913)
Got around to do another and my findings are intriguing here :) 
Go to the tldr(#tldr)

* I am able to observe the QueryOpen on NetCoreClassLibrary right now, even when restore NetCoreConsoleApp. 
* The comments about a regression intrigued me, so I validated that 5.0.303 doesn't have this issue, but 6.0 does. The 5.0.4xx preview version on my box does not repro it either, 5.0.400-preview.21328.4. 
* I then decided to attempt repros with msbuild. All of the behaviors were consistent as dotnet.exe. That means, the SDK being used correlated to whether we have a repro.
* After that, I defined a dummy target, that does absolutely nothing and executed it. Saw the QueryOpen on the project in question. 
* When NuGet restores a project, it needs to evaluate the dependant projects as well, because while NuGet won't restore that project, it needs to know the details about it's framework package references etc. 
* Then I debugged restore and I observed that the QueryOpen happens *before* restore itself is even hit. 

#### TLDR

* The probe is coming at project evaluation, and seems like it's happening regardless of the target being executed. 

I didn't really keep digging to understand where the probe is coming from, but I think SDK folks might be more helpful here. 

cc @dsplaisted Does this sound familiar? Anything that changed in .NET 6 that could lead to a project of the folder?

**Comment by nkolev92** [comment]
Moved here because the access is coming at evaluation time, so likely to be SDK/MSBuild.

**Comment by dsplaisted** [comment]
It looks like it's this change, which checks to see if the assets file exists during evaluation: https://github.com/dotnet/sdk/pull/16840

We might be able to reverse the condition, and then you could prevent the file exist check by setting `EmbedProjectAssetsFile` to false.

**Comment by breidikl** [comment]
That would work, as we could set EmbedProjectAssetsFile to false in our shared logic.  

/bl is not commonly used by our customers, and we have alternatives that support doing that outside of the BuildXL based builds when needed.

**Comment by breidikl** [comment]
I've sent a PR with the change I tested locally at #20612

**Comment by baronfel** [comment]
Closing due to the linked PR having been merged already.
